staging/src/k8s.io/component-base/logs/api/v1/options.go

```diff
@@ -0,0 +1,282 @@
+// can be passed when the struct is not embedded in some larger struct.
+func Validate(c *LoggingConfiguration, featureGate featuregate.FeatureGate, fldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	if c.Format != DefaultLogFormat {
+		// WordSepNormalizeFunc is just a guess. Commands should use it,
+		// but we cannot know for sure.
+		allFlags := unsupportedLoggingFlags(cliflag.WordSepNormalizeFunc)
+		for _, f := range allFlags {
+			if f.DefValue != f.Value.String() {
+				errs = append(errs, field.Invalid(fldPath.Child("format"), c.Format, fmt.Sprintf("Non-default format doesn't honor flag: %s", f.Name)))
+			}
+		}
+	}
+	format, err := logRegistry.get(c.Format)
+	if err != nil {
+		errs = append(errs, field.Invalid(fldPath.Child("format"), c.Format, "Unsupported log format"))
+	} else if format != nil {
+		if format.feature != LoggingStableOptions {
+			enabled := featureGates()[format.feature].Default
+			if featureGate != nil {
+				enabled = featureGate.Enabled(format.feature)
+			}
+			if !enabled {
+				errs = append(errs, field.Forbidden(fldPath.Child("format"), fmt.Sprintf("Log format %s is disabled, see %s feature", c.Format, format.feature)))
+			}
+		}
+	}
+
+	// The type in our struct is uint32, but klog only accepts positive int32.
+	if c.Verbosity > math.MaxInt32 {
+		errs = append(errs, field.Invalid(fldPath.Child("verbosity"), c.Verbosity, fmt.Sprintf("Must be <= %d", math.MaxInt32)))
+	}
+	vmoduleFldPath := fldPath.Child("vmodule")
+	if len(c.VModule) > 0 && c.Format != "" && c.Format != "text" {
+		errs = append(errs, field.Forbidden(vmoduleFldPath, "Only supported for text log format"))
+	}
+	for i, item := range c.VModule {
+		if item.FilePattern == "" {
+			errs = append(errs, field.Required(vmoduleFldPath.Index(i), "File pattern must not be empty"))
+		}
+		if strings.ContainsAny(item.FilePattern, "=,") {
+			errs = append(errs, field.Invalid(vmoduleFldPath.Index(i), item.FilePattern, "File pattern must not contain equal sign or comma"))
+		}
+		if item.Verbosity > math.MaxInt32 {
+			errs = append(errs, field.Invalid(vmoduleFldPath.Index(i), item.Verbosity, fmt.Sprintf("Must be <= %d", math.MaxInt32)))
+		}
+	}
+
+	errs = append(errs, validateFormatOptions(c, featureGate, fldPath.Child("options"))...)
+	return errs
+}
+
+func validateFormatOptions(c *LoggingConfiguration, featureGate featuregate.FeatureGate, fldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	errs = append(errs, validateJSONOptions(c, featureGate, fldPath.Child("json"))...)
+	return errs
+}
+
+func validateJSONOptions(c *LoggingConfiguration, featureGate featuregate.FeatureGate, fldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	if c.Options.JSON.SplitStream && !featureEnabled(featureGate, LoggingAlphaOptions) {
+		errs = append(errs, field.Forbidden(fldPath.Child("splitStream"), fmt.Sprintf("Feature %s is disabled", LoggingAlphaOptions)))
```

Comment:
suggest this format to make sure the lookup and error message use the same level (will prevent accidental drift as we graduate to beta, etc):

```suggestion
	if gate := LoggingAlphaOptions; c.Options.JSON.SplitStream && !featureEnabled(featureGate, gate) {
		errs = append(errs, field.Forbidden(fldPath.Child("splitStream"), fmt.Sprintf("Feature %s is disabled", gate)))
```

Reply:
Good idea, done.